rmw_connext_cpp/src/functions.cpp

```diff
@@ -279,14 +297,45 @@ rmw_create_publisher(
     goto fail;
   }
 
+  // allocates memory, but doesn't have to be freed.
+  // partition operater takes ownership of it.
+  name_tokens = rcutils_split_last(topic_name, '/');
+  if (name_tokens.size == 1) {
+    partition_str = DDS_String_dup(ros_topics_prefix);
+    topic_str = DDS_String_dup(name_tokens.data[0]);
+  } else if (name_tokens.size == 2) {
+    // TODO(Karsten1987): Fix utility function for this
+    size_t partition_length = strlen(ros_topics_prefix) + strlen(name_tokens.data[0]) + 2;
+    char * concat_str = reinterpret_cast<char *>(rmw_allocate(partition_length * sizeof(char)));
+    snprintf(concat_str, partition_length, "%s/%s", ros_topics_prefix, name_tokens.data[0]);
+    // Connext will call deallocate on this, passing ownership to connext
+    partition_str = DDS_String_dup(concat_str);
+    topic_str = DDS_String_dup(name_tokens.data[1]);
```

Comment:
See https://github.com/ros2/rmw_connext/pull/227#discussion_r136463024 It looks like that the variable is not freed and `valgrind` marks it as a leak. Any idea why this was added / for which case?

Reply:
It could be that `partition_str` is actually taking ownership, but `topic_str` is not and needs to be explicitly free'd.

I remember debugging this with @Karsten1987 and he found that Connext was taking ownership of some string, though I don't remember which or both.

Reply:
as far as i remember, the reason why we call DDS_String_dup on it is because connnext takes ownership. We duplicate the string so that we can delete the initial allocated memory (`concat_str`, `name_tokens`)

Reply:
See #244.